spring-kafka/src/test/java/org/springframework/kafka/core/DefaultKafkaConsumerFactoryTests.java

```diff
@@ -59,18 +65,149 @@
 	private EmbeddedKafkaBroker embeddedKafka;
 
 	@Test
-	public void testClientId() {
-		Map<String, Object> configs = Collections.singletonMap(ConsumerConfig.CLIENT_ID_CONFIG, "foo");
-		DefaultKafkaConsumerFactory<String, String> factory = new DefaultKafkaConsumerFactory<String, String>(configs) {
+	public void testProvidedDeserializerInstancesAreShared() {
+		ConsumerFactory<String, String> target = new DefaultKafkaConsumerFactory<>(Collections.emptyMap(), new StringDeserializer() {
 
-			@Override
-			protected KafkaConsumer<String, String> createKafkaConsumer(Map<String, Object> configs) {
-				assertThat(configs.get(ConsumerConfig.CLIENT_ID_CONFIG)).isEqualTo("foo-1");
+		}, null);
+		assertThat(target.getKeyDeserializer()).isSameAs(target.getKeyDeserializer());
+	}
+
+	@Test
+	public void testSupplierProvidedDeserializersAreNotShared() {
+		ConsumerFactory<String, String> target = new DefaultKafkaConsumerFactory<>(Collections.emptyMap(), () -> new StringDeserializer() {
+
+		}, null);
+		assertThat(target.getKeyDeserializer()).isNotSameAs(target.getKeyDeserializer());
+	}
+
+	@Test
+	public void testNoOverrides() {
+		Map<String, Object> originalConfig = Collections.emptyMap();
+		new DefaultKafkaConsumerFactory<String, String>(originalConfig) {
+			protected KafkaConsumer<String, String> createKafkaConsumer(Map<String, Object> configProps) {
+				assertThat(configProps).isEqualTo(originalConfig);
```

Comment:
I don't think this is correct place to perform assertions. We really need to have a code flow like `given, when, then`.

I think you can store a `config` in some `AtomicReference` to reach it in the end of test.

Reply:
Ah, good idea - the target code doesn't lend itself easily to accessing the config that gets passed to the KafkaConsumer and I did feel that my tests looked a bit hacky, so yes, assigning the value to a variable referenced outside the overridden method and asserting after sounds like a better way to go.

At one point I did refactor the DefaultKafkaConsumerFactory code to make it a bit more easy to test this behaviour, but I have not pushed those changes as I don't want to get into any extensive refactoring. I'll just update these tests.

Personally I'm not enjoying going back to JUnit as I've been using Spock for the last few years which is much better for clearly defined tests with explicit given when then clauses!